scenario/mocking.py

```diff
@@ -784,21 +796,79 @@ def _find_exec_handler(self, command) -> Optional["Exec"]:
         # matter how much of it was used, so we have failed to find a handler.
         return None
 
-    def exec(self, command, **kwargs):  # noqa: U100 type: ignore
+    def exec(
+        self,
+        command: List[str],
+        *,
+        environment: Optional[Dict[str, str]] = None,
+        working_dir: Optional[str] = None,
+        timeout: Optional[float] = None,
+        user_id: Optional[int] = None,
+        user: Optional[str] = None,
+        group_id: Optional[int] = None,
+        group: Optional[str] = None,
+        stdin: Optional[Union[str, bytes, TextIO]] = None,
+        stdout: Optional[TextIO] = None,
+        stderr: Optional[TextIO] = None,
+        encoding: Optional[str] = "utf-8",
+        combine_stderr: bool = False,
+        **kwargs,
+    ):
         handler = self._find_exec_handler(command)
         if not handler:
-            raise RuntimeError(
-                f"mock for cmd {command} not found. Please pass to the Container "
-                f"{self._container.name} a scenario.Exec mock for the "
-                f"command your charm is attempting to run, or patch "
-                f"out whatever leads to the call.",
+            raise ExecError(
+                command,
+                127,
+                "",
+                f"mock for cmd {command} not found. Please patch out whatever "
+                f"leads to the call, or pass to the Container {self._container.name} "
+                f"a scenario.Exec mock for the command your charm is attempting "
+                f"to run, such as "
+                f"'Container(..., execs={{scenario.Exec({list(command)}, ...)}})'",
             )
 
+        if stdin is None:
+            proc_stdin = self._transform_exec_handler_output("", encoding)
+        else:
+            proc_stdin = None
+            stdin = stdin.read() if hasattr(stdin, "read") else stdin  # type: ignore
+        if stdout is None:
+            proc_stdout = self._transform_exec_handler_output(handler.stdout, encoding)
+        else:
+            proc_stdout = None
+            stdout.write(handler.stdout)
+        if stderr is None:
+            proc_stderr = self._transform_exec_handler_output(handler.stderr, encoding)
+        else:
+            proc_stderr = None
+            stderr.write(handler.stderr)
+
+        args = ExecArgs(
```

Comment:
Should the following use keyword args? The order doesn't seem at all obvious right now.

Reply:
Isn't that more an argument that `ops.testing.ExecArgs` should make those arguments keyword-only? :smile: 

Sure, I can adjust them to be keyword.